src/librustc_mir/interpret/cast.rs

```diff
@@ -12,25 +11,36 @@ use rustc_middle::ty::{self, Ty, TypeAndMut, TypeFoldable};
 use rustc_span::symbol::sym;
 use rustc_target::abi::{LayoutOf, Size, Variants};
 
+use super::{truncate, FnVal, ImmTy, Immediate, InterpCx, Machine, OpTy, PlaceTy};
+
 impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
     pub fn cast(
         &mut self,
         src: OpTy<'tcx, M::PointerTag>,
-        kind: CastKind,
+        cast_kind: CastKind,
+        cast_ty: Ty<'tcx>,
         dest: PlaceTy<'tcx, M::PointerTag>,
     ) -> InterpResult<'tcx> {
         use rustc_middle::mir::CastKind::*;
-        match kind {
+        // FIXME: In which cases should we trigger UB when the source is uninit?
+        match cast_kind {
             Pointer(PointerCast::Unsize) => {
-                self.unsize_into(src, dest)?;
+                let cast_ty = self.layout_of(cast_ty)?;
+                self.unsize_into(src, cast_ty, dest)?;
             }
 
-            Misc | Pointer(PointerCast::MutToConstPointer | PointerCast::ArrayToPointer) => {
+            Misc => {
                 let src = self.read_immediate(src)?;
-                let res = self.cast_immediate(src, dest.layout)?;
+                let res = self.misc_cast(src, cast_ty)?;
                 self.write_immediate(res, dest)?;
             }
 
+            Pointer(PointerCast::MutToConstPointer | PointerCast::ArrayToPointer) => {
```

Comment:
`ArrayToPointer` is confusing, I'm guessing `ArrayToElemPointer` makes more sense (or should it be `ArayRefToElemPointer`? since a regular raw pointer cast shouldn't care about the pointee type changing).

Although `PointerCast` variants might not need the `Pointer` suffix. Anyway, this is not that important to this PR.

Reply:
Yeah well I never understood most of these (and at some point whoever designed this clearly gave up and stuffed everything else into `Misc`).